heft-plugins/heft-rspack-plugin/src/RspackPlugin.ts

```diff
@@ -0,0 +1,489 @@
+  ): Promise<TRspack.Compiler | TRspack.MultiCompiler> {
+    if (!this._rspackCompiler) {
+      const rspack: RspackCoreImport = await this._loadRspackAsync();
+      taskSession.logger.terminal.writeLine(`Using Rspack version ${rspack.version}`);
+      this._rspackCompiler = Array.isArray(rspackConfiguration)
+        ? rspack.default(rspackConfiguration) /* (rspack.Compilation[]) => MultiCompiler */
+        : rspack.default(rspackConfiguration); /* (rspack.Compilation) => Compiler */
+    }
+    return this._rspackCompiler;
+  }
+
+  private async _runRspackAsync(
+    taskSession: IHeftTaskSession,
+    heftConfiguration: HeftConfiguration,
+    options: IRspackPluginOptions
+  ): Promise<void> {
+    this._validateEnvironmentVariable(taskSession);
+    if (taskSession.parameters.watch || this._isServeMode) {
+      // Should never happen, but just in case
+      throw new InternalError('Cannot run Rspack in compilation mode when watch mode is enabled');
+    }
+
+    // Load the config and compiler, and return if there is no config found
+    const rspackConfiguration: IRspackConfiguration | undefined = await this._getRspackConfigurationAsync(
+      taskSession,
+      heftConfiguration,
+      options
+    );
+    if (!rspackConfiguration) {
+      return;
+    }
+    const compiler: TRspack.Compiler | TRspack.MultiCompiler = await this._getRspackCompilerAsync(
+      taskSession,
+      rspackConfiguration
+    );
+    taskSession.logger.terminal.writeLine('Running Rspack compilation');
+
+    // Run the rspack compiler
+    let stats: TRspack.Stats | TRspack.MultiStats | undefined;
+    try {
+      stats = await LegacyAdapters.convertCallbackToPromise(
+        (compiler as TRspack.Compiler).run.bind(compiler)
+      );
+      await LegacyAdapters.convertCallbackToPromise(compiler.close.bind(compiler));
+    } catch (e) {
+      taskSession.logger.emitError(e as Error);
+    }
+
+    // Emit the errors from the stats object, if present
+    if (stats) {
+      this._recordErrors(stats, heftConfiguration.buildFolderPath);
+      this._emitErrors(taskSession.logger);
+      if (this.accessor.hooks.onEmitStats.isUsed()) {
+        await this.accessor.hooks.onEmitStats.promise(stats);
+      }
+    }
+  }
+
+  private async _runRspackWatchAsync(
+    taskSession: IHeftTaskSession,
+    heftConfiguration: HeftConfiguration,
+    options: IRspackPluginOptions,
+    requestRun: () => void
+  ): Promise<void> {
+    // Save a handle to the original promise, since the this-scoped promise will be replaced whenever
+    // the compilation completes.
+    let rspackCompilationDonePromise: Promise<void> | undefined = this._rspackCompilationDonePromise;
+
+    if (!this._rspackCompiler) {
+      this._validateEnvironmentVariable(taskSession);
+      if (!taskSession.parameters.watch) {
+        // Should never happen, but just in case
+        throw new InternalError('Cannot run Rspack in watch mode when compilation mode is enabled');
```

Comment:
Message is backwards

Reply:
updated message here and in the webpack5 plugin